src/workerd/jsg/fast-api.h

```diff
@@ -0,0 +1,131 @@
+};
+
+template <typename U>
+struct IsFastApiMaybe<jsg::LenientOptional<U>> {
+  static constexpr bool value = IsFastApiCompatibleWrappedType1<U>::value;
+};
+
+template <typename T>
+struct IsFastApiCompatibleWrappedType {
+  static constexpr bool value =
+      IsFastApiCompatibleWrappedType1<T>::value || IsFastApiMaybe<T>::value;
+};
+
+// Helper to determine if a type can be used as a return type in V8 Fast API
+template <typename T>
+struct IsFastApiCompatibleReturnType {
+  static constexpr bool value = kj::isSameType<T, void>() || kj::isSameType<T, bool>() ||
+      kj::isSameType<T, int32_t>() || kj::isSameType<T, uint32_t>() || kj::isSameType<T, float>() ||
+      kj::isSameType<T, double>();
+};
+
+// Helper to determine if a type can be used as a parameter in V8 Fast API
+template <typename T>
+struct IsFastApiCompatibleParamType {
+  static constexpr bool value = IsFastApiCompatibleReturnType<T>::value ||
+      kj::isSameType<T, v8::Local<v8::Value>>() || kj::isSameType<T, v8::Local<v8::Object>>() ||
+      IsFastApiCompatibleWrappedType<T>::value;
+};
+
+// Specialization for jsg::Lock& - this is compatible with Fast API.
+// This allows methods that take jsg::Lock& as their first parameter to be used with V8 Fast API.
+// The Lock will be extracted from the V8 isolate in the fast path, enabling methods that need
+// access to the V8 context and isolate to still benefit from the performance of Fast API.
+template <>
+struct IsFastApiCompatibleParamType<jsg::Lock&> {
+  static constexpr bool value = true;
+};
+
+// Helper to determine if a method can be used with V8 fast API
+template <typename Ret, typename... Args>
+struct IsFastApiCompatible {
+  static constexpr bool value = IsFastApiCompatibleReturnType<Ret>::value &&
+      (IsFastApiCompatibleParamType<Args>::value && ...);
+};
+
+// Helper to determine if a method pointer type is compatible with Fast API
+template <typename Method>
+struct IsFastMethodCompatible {
+  static constexpr bool value = false;
+};
+
+// Specialization for non-static member functions
+template <typename Class, typename Ret, typename... Args>
+struct IsFastMethodCompatible<Ret (Class::*)(Args...)> {
+  static constexpr bool value = IsFastApiCompatible<Ret, Args...>::value;
+};
+
+// Specialization for const non-static member functions
+template <typename Class, typename Ret, typename... Args>
+struct IsFastMethodCompatible<Ret (Class::*)(Args...) const> {
+  static constexpr bool value = IsFastApiCompatible<Ret, Args...>::value;
+};
+
+// Implements V8 Fast API callback for a method. This class handles both the fast and slow paths.
+template <typename TypeWrapper,
+    const char* methodName,
+    bool isContext,
+    typename T,
+    typename Method,
+    Method method,
+    typename Indexes>
+struct FastMethodCallback;
+
+template <typename T, typename Enable = void>
+struct FastApiJSGToV8 {
+  static_assert(sizeof(T) != sizeof(T), "Type is not supported in fast API");
```

Comment:
When would `sizeof(T)`  not equal `sizeof(T)`? A comment would be helpful because I wouldn't know.